easydock/preparation_for_docking.py

```diff
@@ -35,43 +38,246 @@ def mol_from_smi_or_molblock(ligand_string):
+        for ring in ssr:
+            is_atom_saturated_array = [mol.GetAtomWithIdx(atom_id).GetHybridization() == Chem.HybridizationType.SP3 for atom_id in ring]
+            if all(is_atom_saturated_array):
+                saturated_ring_list.append(ring)
+        return saturated_ring_list
+
+    def gen_conf(mole: Chem.Mol, useRandomCoords: bool, randomSeed: int, has_saturated_ring: bool) -> tuple[Chem.Mol, float]:
         params = AllChem.ETKDGv3()
         params.useRandomCoords = useRandomCoords
         params.randomSeed = randomSeed
-        conf_stat = AllChem.EmbedMolecule(mole, params)
+        if has_saturated_ring:
+            #10 is used as default according to the C language documentation iirc, but I have to specify the numbers.
+            conf_stat = AllChem.EmbedMultipleConfs(mole, 10, params)
+        else:
+            conf_stat = AllChem.EmbedMolecule(mole, params)
         return mole, conf_stat
-
+    
+    def remove_confs_rms(mol: Chem.Mol, saturated_ring_list: list[list[int]], rms: float=0.25, keep_nconf: Optional[int]=None) -> Chem.Mol:
+        """
+        The function uses AgglomerativeClustering to select conformers.
+
+        :param mol: input molecule with multiple conformers
+        :param rms: discard conformers which are closer than given value to a kept conformer
+        :param keep_nconf: keep at most the given number of conformers. This parameter has precedence over rms
+        :return:
+        """
+
+        def gen_ids(ids: int) -> Iterator[int]:
+            for i in range(1, len(ids)):
+                for j in range(0, i):
+                    yield j, i
+
+        if keep_nconf and mol.GetNumConformers() <= keep_nconf:
+            return mol
+
+        if mol.GetNumConformers() <= 1:
+            return mol
+
+        mol_tmp = Chem.RemoveHs(mol)   # calc rms for heavy atoms only
+        rms_ = GetConformerRMSMatrixForSaturatedRingMolecule(mol_tmp, atomIds=saturated_ring_list, prealigned=False)
+
+        cids = [c.GetId() for c in mol_tmp.GetConformers()]
+        arr = np.zeros((len(cids), len(cids)))
+        for (i, j), v in zip(gen_ids(cids), rms_):
+            arr[i, j] = v
+            arr[j, i] = v
+
+        cl = AgglomerativeClustering(n_clusters=None, linkage='complete', metric='precomputed', distance_threshold=rms).fit(arr)
+
+        keep_ids = []
+        for i in set(cl.labels_):
+            ids = np.where(cl.labels_ == i)[0]
+            j = arr[np.ix_(ids, ids)].mean(axis=0).argmin()
+            keep_ids.append(cids[j])
+        remove_ids = set(cids) - set(keep_ids)
+
+        for cid in sorted(remove_ids, reverse=True):
+            mol.RemoveConformer(cid)
+
+        if mol.GetNumConformers() == 1:
+            return mol
+        
+        if keep_nconf:
+            if mol.GetNumConformers() <= keep_nconf:
+                return mol
+            
+            cids = [c.GetId() for c in mol.GetConformers()]
+            arr = arr[np.ix_(cids, cids)]
```

Comment:
This is a potential issue. Conformer ids are arbitrary, they do not guarantee to be sequential from 0 to N-1 as atoms in RDKit. Thus, to be safe, another indexing should be used, probably involving `keep_ids`.

Reply:
so, instead of `cids`, we can use `keep_ids` as the reference to be removed for the nkeep filters? Something like this?

```python
            arr = arr[np.ix_(keep_ids, keep_ids)]
            keep_ids_nconf_filter = []
            #calculation 
            remove_ids = set(keep_ids) - set(keep_ids_nconf_filter)
```